ui/src/routes/incarnations/Form.tsx

```diff
@@ -166,35 +182,35 @@ export const IncarnationsForm = ({
             )} />
         </Hug>
       )}
-      <h4>Template data</h4>
-      <Hug mb={16}>
-        {fields.map((field, index) => (
-          <Hug flex mb={8} mx={-4} key={field.id}>
-            <Hug w="50%" px={4}>
-              <TextField disabled={isLoading} placeholder="Key" {...register(`templateData.${index}.key` as const)} />
-            </Hug>
-            <Hug w="50%" pl={4} pr={8}>
-              <TextField disabled={isLoading} placeholder="Value" {...register(`templateData.${index}.value` as const)} />
-            </Hug>
-            <Hug pr={4}>
-              <IconButton disabled={isLoading} type="button" onClick={() => remove(index)} title="Delete">
-                <Trash />
-              </IconButton>
-            </Hug>
-          </Hug>
-        ))}
-      </Hug>
-      <Hug flex mb={8} ml={-8}>
-        <Hug w="calc(50% - 18px)" pl={8} pr={4}>
-          <TextField disabled={isLoading} placeholder="Start typing to add new key..." value="" onChange={e => {
-            appendAndFocus(e.target.value, '')
-          }} />
-        </Hug>
-        <Hug w="calc(50% + 18px)" pl={4}>
-          <TextField disabled={isLoading} placeholder="...value pair" value="" onChange={e => {
-            appendAndFocus('', e.target.value)
-          }} />
-        </Hug>
+      <Hug my={16}>
+        {isEdit ? (
+          <Tabs
+            tabs={[
+              {
+                label: <strong>Template data JSON</strong>,
+                content: <Hug my={8}>
+                  {editTemplateDataController}
+                </Hug>
+              },
+              {
+                label: <strong>Full template data (readonly)</strong>,
+                content: <Hug my={8}>
+                  <JsonEditor
+                    defaultValue={JSON.stringify(templateDataFull, null, 2)}
+                    readOnly
+                    height="max(calc(100vh - 700px), 200px)" />
```

Comment:
For readability why not move this into a variable like `editTemplateDataController`

Reply:
sorry, I didn't get it, the editor is already in this variable